### ISSUE TITLE
etcd v3 gets overloaded when db is reaching 600MB with large keys

### ISSUE THREAD

**Issue by gyuho** [coreos/etcd#5245]
```
PUT
key:   time.Now.Unix
value: http://www.gutenberg.org/cache/epub/1400/pg1400.txt  # 1MB text data
```

```
# leader log with about 500 put requests
2016-05-02 16:51:12.810725 W | etcdserver: failed to send out heartbeat on time (deadline exceeded for 67.458382ms)
2016-05-02 16:51:12.810778 W | etcdserver: server is likely overloaded

# data.etcd/member/snap/db is about 570MB
```

<br><hr>
#### Code

```go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"time"

	"golang.org/x/net/context"

	"github.com/coreos/etcd/clientv3"
	"github.com/dustin/go-humanize"
)

/*
go get -v -u golang.org/x/net/context
go get -v -u github.com/coreos/etcd/clientv3
go get -v -u github.com/dustin/go-humanize
*/
func main() {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   []string{"10.240.0.17:2379", "10.240.0.18:2379", "10.240.0.19:2379"},
		DialTimeout: 3 * time.Second,
	})
	if err != nil {
		log.Fatal(err)
	}
	defer cli.Close()

	// http://www.gutenberg.org/cache/epub/1400/pg1400.txt
	bts, err := ioutil.ReadFile("a.txt")
	if err != nil {
		log.Fatal(err)
	}
	value := string(bts)
	log.Println("Writing", humanize.Bytes(uint64(len(bts))))

	for i := 0; i < 700; i++ {
		key := fmt.Sprintf("%d", time.Now().Unix())
		stp := time.Now()
		_, err = cli.Put(context.TODO(), key, value)
		if err != nil {
			log.Fatal(err)
		}
		log.Printf("%d: etcd Put took %v", i, time.Since(stp))

		// time.Sleep(100 * time.Millisecond)
	}
}
```

**Comment by xiang90** [comment]
@gyuho I think here we are putting really large values. Does this happen if you put small values? I think I can put more than 5gb small keys into etcd without causing any performance degradation.

**Comment by gyuho** [comment]
@xiang90 I get this with around 300kb keys as well (when the db is around 600MB). I will try smaller keys.

**Comment by gyuho** [comment]
I tried putting 200kb values and don't see this behavior. This only happens with large values (>300kb).

**Comment by gyuho** [comment]
Did 1MB key writes with 3-node cluster with GCE VM, 8 vCPUs + 16GB Memory + 50GB SSD. And when the `TickMS` (heartbeat interval) is 100ms and the client keeps writing 1MB keys, we observed that such fsync-s take more than usual, causing `etcdserver: server is likely overloaded` warnings.

```
[DEBUG] t.tx.Commit took: 272.601775ms
[DEBUG] WAL sync took: 254.31782ms
2016-05-11 03:53:18.437574 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 89.421358ms)
2016-05-11 03:53:18.437612 W | etcdserver: server is likely overloaded
2016-05-11 03:53:18.437624 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 89.474918ms)
2016-05-11 03:53:18.437633 W | etcdserver: server is likely overloaded
```

The testing machine has 50GB SSD, and this [document](https://cloud.google.com/compute/docs/disks/performance) explains that GCE remote SSD has READ/WRITE throughput of `0.48 MB/s` per GB, which means the machine has READ/WRITE throughput of `25 MB/s`.

And the client was writing in the rate of `30 MB/s`, **which is already beyond the limit**. This slows down when it has written about >500 MB data, and that's when the etcd logs complain about `etcdserver: server is likely overloaded`. We can confirm this by measuring fsync duration.

For this use case, etcd does disk-writes when

1. Writing data onto [WAL file](https://github.com/coreos/etcd/blob/master/wal/wal.go#L488-L515)
2. In-memory transaction gets [committed](https://github.com/coreos/etcd/blob/master/mvcc/backend/batch_tx.go#L158-L168)

`server is likely overloaded` happened when the WAL and batch commit operation took +200ms, which is greater than heartbeat interval 100ms.

To summarize, it's not an etcd problem. If anyone comes across this issue, please check your workload on etcd and disk performance.
